src/Microsoft.OData.Client/Materialization/ODataEntityMaterializer.cs

```diff
@@ -803,6 +803,56 @@ internal object ProjectionValueForPath(MaterializerEntry entry, Type expectedTyp
             return result;
         }
 
+        /// <summary>Projects a simple dynamic value from the specified <paramref name="path"/>.</summary>
+        /// <param name="entry">Root entry for paths.</param>
+        /// <param name="expectedPropertyType">Expected type for the dynamic property value.</param>
+        /// <param name="path">Path to pull value for.</param>
+        /// <returns>The value for the specified <paramref name="path"/>.</returns>
+        /// <remarks>
+        /// This method will not instantiate entity types, except to satisfy requests
+        /// for payload-driven feeds or leaf entities.
+        /// </remarks>
+        internal object ProjectionDynamicValueForPath(MaterializerEntry entry, Type expectedPropertyType, ProjectionPath path)

```

Comment:
@habbes Say you have `Sale` entity containing 3 fields: Id, ProductId, and Amount. On the client, the class definition will be:
```csharp
public class Sale
{
    public int Id { get; set; }
    public string ProductId { get; set; }
    public decimal Amount { get; set; }
}
```
If you want to return the average amount for each product, you'd use the following OData URL:
```
/Sales?$apply=groupby((ProductId),aggregate(Amount with average as AverageAmount))
```
The response would look as follows:
```json
{
    "@odata.context": "http://tempuri.org/$metadata#Sales(ProductId, AverageAmount)",
    "value": [{
        "@odata.id": null,
        "ProductId": "P1",
        "AverageAmount": 16.00
    }]
}
```
Because of the use of aliases when performing aggregation, the response contains a dynamic property `AverageAmount` that is not defined on the `Sale` entity. The `ProjectionValueForPath` method was designed to read value for declared properties - specifically for `$select` projections. I added `ProjectionDynamicValueForPath` to read dynamic properties like `AverageAmount` above.